rir/src/compiler/opt/force_dominance.cpp

```diff
@@ -136,15 +133,63 @@ struct ForcedBy {
             }
         }
 
-        if (eagerFunction != NotEagerFunction &&
-            eagerFunction != other.eagerFunction) {
-            eagerFunction = NotEagerFunction;
+        if (forceArgsBeforeEffect && !other.forceArgsBeforeEffect) {
+            forceArgsBeforeEffect = false;
+            res.update();
+        }
+
+        if (!ambiguousForceOrder && other.ambiguousForceOrder) {
+            ambiguousForceOrder = true;
             res.update();
         }
 
+        if (argumentForceOrder != other.argumentForceOrder) {
+            auto mySize = argumentForceOrder.size();
+            auto otherSize = other.argumentForceOrder.size();
+            auto common = mySize;
+
+            if (mySize > otherSize) {
+                argumentForceOrder.resize(otherSize);
+                ambiguousForceOrder = true;
+                common = otherSize;
+                res.update();
+            } else if (!ambiguousForceOrder && otherSize > mySize) {
+                ambiguousForceOrder = true;
+                res.update();
+            }
+
+            for (size_t i = 0; i < common; ++i) {
+                if (argumentForceOrder[i] != other.argumentForceOrder[i]) {
+                    argumentForceOrder.resize(i);
+                    ambiguousForceOrder = true;
+                    res.update();
+                    break;
+                }
+            }
+        }
+
         return res;
     }
 
+    bool maybeForced(size_t i) const {
+        for (auto f : argumentForceOrder) {
+            if (f == i)
+                return true;
+        }
+        return ambiguousForceOrder;
+    }
+
+    bool eagerLikeFunction(ClosureVersion* fun) const {
+        if (!forceArgsBeforeEffect)
+            return false;
+        if (ambiguousForceOrder || argumentForceOrder.size() < fun->nargs())
+            return false;
+        for (size_t i = 0; i < fun->nargs(); ++i)
+            if (argumentForceOrder[i] != i)
```

Comment:
I'm trying to figure out what is going on here, and it seems like `eagerLikeFunction` requires _all_ arguments to be forced, in left-to-right order?

So if I had a function

```
f <- function(a1, a2, a3, a4) ...
```

the arguments need to be forced in the order `a1`, `a2`, `a3`, and `a4`?

And if they're force in some other order, e.g., `a2`, `a1`, `a3`, `a4`, that's no good?

Reply:
yes